src/server/Config.py

```diff
@@ -1,131 +1,256 @@
-# override defaults above with config from file
-try:
-    with open(CONFIG_FILE_NAME, 'r') as f:
-        ExternalConfig = json.load(f)
-
-    GENERAL.update(ExternalConfig['GENERAL'])
-
-    # if auth fields set to empty strings then allow open access to all pages
-    if ExternalConfig['GENERAL'].get('ADMIN_USERNAME') == "" and \
-                ExternalConfig['GENERAL'].get('ADMIN_PASSWORD') == "":
-        GENERAL['ADMIN_USERNAME'] = ''
-        GENERAL['ADMIN_PASSWORD'] = ''
-
-    if 'HARDWARE' in ExternalConfig:
-        HARDWARE.update(ExternalConfig['HARDWARE'])
-    if 'LOGGING' in ExternalConfig:
-        LOGGING.update(ExternalConfig['LOGGING'])
-    if 'LED' in ExternalConfig:
-        LED.update(ExternalConfig['LED'])
-    if 'VRX_CONTROL' in ExternalConfig:
-        VRX_CONTROL.update(ExternalConfig['VRX_CONTROL'])
-
-
-    '''  #pylint: disable=pointless-string-statement
-    # Subtree updating
-    try:
-        bitmaptree = LED['BITMAPS']
-        LED'].update(ExternalLED'])
-        LED['BITMAPS'] = bitmaptree
-        LED['BITMAPS'].update(ExternalLED['BITMAPS'])
-    except KeyError:
-        if 'LED' in ExternalConfig:
-            LED'].update(ExternalLED'])
-        else:
-            print "No 'LED' entry found in configuration file "
-    '''
-
-    if 'SENSORS' in ExternalConfig:
-        SENSORS.update(ExternalConfig['SENSORS'])
-    if 'SERIAL_PORTS' in ExternalConfig:
-        SERIAL_PORTS.extend(ExternalConfig['SERIAL_PORTS'])
-    GENERAL['configFile'] = 1
-    InitResultStr = "Using configuration file '{0}'".format(CONFIG_FILE_NAME)
-    InitResultLogLevel = logging.INFO
-except IOError:
-    GENERAL['configFile'] = 0
-    InitResultStr = "No configuration file found, using defaults"
-    InitResultLogLevel = logging.WARN
-except ValueError as ex:
-    GENERAL['configFile'] = -1
-    InitResultStr = "Configuration file invalid, using defaults; error is: " + str(ex)
-    InitResultLogLevel = logging.ERROR
-
-# Apply legacy config options for backward compatibility
-if not GENERAL['SECONDARIES']:
-    if 'SLAVES' in GENERAL and GENERAL['SLAVES']:
-        GENERAL['SECONDARIES'] = GENERAL['SLAVES']
-
-if GENERAL['SECONDARY_TIMEOUT'] == 300:
-    if 'SLAVE_TIMEOUT' in GENERAL and GENERAL['SLAVE_TIMEOUT']:
-        GENERAL['SECONDARY_TIMEOUT'] = GENERAL['SLAVE_TIMEOUT']
-
-if LED['LED_GPIO'] == 10:
-    if 'LED_PIN' in LED and LED['LED_PIN']:
-        LED['LED_GPIO'] = LED['LED_PIN']
-
-# Writes a log message describing the result of the module initialization.
-def logInitResultMessage():
-    if InitResultStr:
-        logger.log(InitResultLogLevel, InitResultStr)
+class Config:
+    def __init__(self, racecontext, filename='config.json'):
+        self._racecontext = racecontext
+        self.filename = filename
+
+        self.config = {
+            'SECRETS': {},
+            'GENERAL': {},
+            'HARDWARE': {},
+            'LED': {},
+            'LOGGING': {},
+            'SENSORS': {},
+        }
+
```

Comment:
Defaults for `ADMIN_USERNAME` and `ADMIN_PASSWORD`

 `ADMIN_USERNAME` and `ADMIN_PASSWORD` are currently only generated in the new config when migrating from the legacy config. New installations cannot log into the timer without manually adding the parameters to the config file.